ggml/src/ggml-rpc/ggml-rpc.cpp

```
@@ -461,10 +475,17 @@ static rpc_tensor serialize_tensor(const ggml_tensor * tensor) {
 }
 
 static void ggml_backend_rpc_buffer_init_tensor(ggml_backend_buffer_t buffer, ggml_tensor * tensor) {
-    UNUSED(buffer);
-    if (ggml_is_quantized(tensor->type)) {
-        // TODO: this check is due to MATRIX_ROW_PADDING in CUDA and should be generalized
-        GGML_ASSERT(tensor->ne[0] % 512 == 0 && "unsupported quantized tensor");
+    ggml_backend_rpc_buffer_context * ctx = (ggml_backend_rpc_buffer_context *)buffer->context;
+
+    // CUDA backend on the server pads everything to 512 due to CUDA limitations.
+    // Due to bandwidth constraints, we only call the server init tensor functions if necessary.
+    if (ggml_is_quantized(tensor->type) && (tensor->ne[0] % 512 != 0)) {
+        rpc_msg_init_tensor_req request;
+
+        request.tensor = serialize_tensor(tensor);
```

What happens here if the tensor is a view? Is the view parent serialized as well?

---

It may happen when using KV quantization, but in case it should be handled correctly, or it may lead to silent failures in the future.

---

Additionally, `init_tensor` may modify the `tensor->extra` value, so to forward this call correctly, it must also return the modified tensor from the server, and update it on the client.

---

Ah ok! Checking with quantized KV does require forwarding ``view_src`` tensors as well, I have added some functionality to send the base tensors in this case. Unfortunately this makes it very slow with quantized KV -- a CUDA specific optimization would be to just forward even less, and not bother forwarding tensor views at all, as currently the CUDA code only clears padding for base tensors. From a quick look, this optimization seems ok with the SYCL, Vulkan and Metal backends too. Do you think this optimization is safe enough in general?

For the ``tensor->extra`` field, I would prefer to just check for that and raise an error at this stage, I'm not sure of an elegant way to implement that.